### ISSUE TITLE
Literal narrowing does not work when wrapped inside a class

### ISSUE THREAD

**Issue by michelbl** [microsoft/pyright#6646]
Type narrowing is a really useful pattern, notably when the narrowed type is a Literal, and the type guard is of the form `if v in ("v1", "v2"):`.

Unfortunately, it does not work when the narrowed value is a member of a class:

```python
from typing import Literal

MyLiteral = Literal["A", "B"]


def use_value(v: bool) -> None:
    pass


def process_literal(literal: MyLiteral):
    if literal in ("A", "B"):
        not_unbound = False

    use_value(not_unbound)  # no error, as it should be 😌


class LiteralWrapper:
    def __init__(self, my_literal: MyLiteral):
        self.my_literal = my_literal


def process_class_1(c: LiteralWrapper):
    if c.my_literal in ("A", "B"):
        not_unbound = False

    use_value(
        not_unbound
    )  # error: "not_unbound" is possibly unbound (reportUnboundVariable)


def process_class_2(c: LiteralWrapper):
    my_literal = c.my_literal
    if my_literal in ("A", "B"):
        not_unbound = False

    use_value(
        not_unbound
    )  # error: "not_unbound" is possibly unbound (reportUnboundVariable)
```

Tested using pyright CLI 1.1.338.

**Comment by erictraut** [comment]
This code relies on an advanced type checking feature that I call [narrowing for implied else](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=narrowing-for-implied-else). It was very difficult to implement and very costly at evaluation time. For those reasons, I needed to impose limits on how it works. Refer to the documentation for these limits.

Lifting these limits would result in a significant performance hit when analyzing typical Python code, so it wouldn't be a good tradeoff to do so.

You can make this work by making some minor changes in your code.

```python
class LiteralWrapper:
    def __init__(self, my_literal: MyLiteral):
        # Add a type annotation here.
        self.my_literal: MyLiteral = my_literal


def process_class_1(c: LiteralWrapper):
    # Add a local variable here.
    my_literal: MyLiteral = c.my_literal
    if my_literal in ("A", "B"):
        not_unbound = False

    use_value(not_unbound)
```

**Comment by michelbl** [comment]
Thank you for your quick and instructive reply!

I'm sorry to have reported a behavior that was indeed documented.
